### PR TITLE
Outbound stats remastered

### PR DESCRIPTION
**Pull request by kriskowal**
This is intended to replace #561. Rebased atop StatEmitter interface.

### CODE REVIEW

**Comment by Raynos** [comment]
@kriskowal what about the common tags?

We should test the stat emitter to make sure it emits properly structured stats.

**Comment by kriskowal** [comment]
The common tags code looks like it will work without modification. We agreed to name our method emitStat and it does the same thing. Yours overloads StatEmitter and does some additional work.

File: node/channel.js [hunk 1]
```diff
@@ -439,4 +450,22 @@ TChannel.prototype.close = function close(callback) {
...
+
+    var commonTags = self.statTags;
+    var commonKeys = Object.keys(self.statTags);
+
+    var localTags = stat.tags;
+    for (var i = 0; i < commonKeys.length; i++) {
+        localTags[commonKeys[i]] = commonTags[commonKeys[i]];
+    }
+
+    self.statEvent.emit(stat);
```
Comment (Raynos) [inline node/channel.js]: I think i broke the `emit()` interface here and it should be `emit(self, stat)`

**Comment by Raynos** [comment]
One of us should add a test for this stuff then ship it.

**Comment by Raynos** [comment]
@kriskowal @jcorbin I took this branch over and it's ready for review.

@mmihic I added two commits to this branch to document my addition to metrics; please check them out and sign them off :)

File: docs/metrics.md [hunk 2]
```diff
@@ -84,6 +84,13 @@ calls initiated by this service to the target service/endpoint.  MUST include
 an additional ``type`` tag indicating the type of error received; valid values are
 the error code names described in the TChannel protocol doc.
 
+##### outbound.calls.operational-errors
+The total number of errors that occured whilst sending outgoing calls initiated
+by this service to the target service/endpoint. These are not error frames but
+actual local socket level errors or timeout errors. SHOULD include an additional
+``type`` tag indicating the type of error received; Standardizing operational
```
Comment (jcorbin) [inline docs/metrics.md]: "; Stan" => "; stan"

File: docs/metrics.md [hunk 3]
```diff
@@ -120,6 +127,9 @@ Latency of an individual attempt (initial request or retry) at a call.
 Measured from the time the first request fragment write is initiated to the
 time the last response fragment read is received.  SHOULD include the peer tag
 containing the host and port of the peer to whom the call attempt was sent.
+SHOULD include a ``retry-count`` tag indicating the per-request retry count
+represented by the metric (e.g. the first outgoing attempt should have 
+retry-count=0, the second outgoing attempt should have retry-count=2 etc)
```
Comment (jcorbin) [inline docs/metrics.md]: let's code-quote each "retry-count=N"

**Comment by jcorbin** [comment]
lgtm, please use injected time rather than Date.now

File: node/test/request-stats.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        app: 'client',
+        host: os.hostname()
+    };
+    client.makeSubChannel({
+        serviceName: 'server',
+        peers: [server.hostPort]
+    });
+
+    client.on('stat', function onStat(stat) {
+        stats.push(stat);
```
Comment (kriskowal) [inline node/test/request-stats.js]: This is pretty great.

**Comment by kriskowal** [comment]
:ok:

**Comment by mmihic** [comment]
lgtm on the metric changes

**Pull request merged**